mem-efficient learnable fake quantization

haichuan-fb opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49315 mem-efficient learnable fake quantization**

The updated quantization operator have a different gradient calculation for scale and zero_point when the output is at the endpoints of clamp operation. The updated quantization operator calculates the gradient according to the gradient of the `clamp` function. This behavior is consistent with the gradient calculation for non-learnable fake-quant ops.

Differential Revision: [D24479735](https://our.internmc.facebook.com/intern/diff/D24479735/)

Commit:
mem-efficient learnable fake quantization

The updated quantization operator have a different gradient calculation for scale and zero_point when the output is at the endpoints of clamp operation. The updated quantization operator calculates the gradient according to the gradient of the `clamp` function. This behavior is consistent with the gradient calculation for non-learnable fake-quant ops.

Differential Revision: [D24479735](https://our.internmc.facebook.com/intern/diff/D24479735/)

[ghstack-poisoned]

Commit:
Update on "mem-efficient learnable fake quantization"


The updated quantization operator have a different gradient calculation for scale and zero_point when the output is at the endpoints of clamp operation. The updated quantization operator calculates the gradient according to the gradient of the `clamp` function. This behavior is consistent with the gradient calculation for non-learnable fake-quant ops.

Differential Revision: [D24479735](https://our.internmc.facebook.com/intern/diff/D24479735/)

[ghstack-poisoned]

Commit:
Update on "mem-efficient learnable fake quantization"


The updated quantization operator have a different gradient calculation for scale and zero_point when the output is at the endpoints of clamp operation. The updated quantization operator calculates the gradient according to the gradient of the `clamp` function. This behavior is consistent with the gradient calculation for non-learnable fake-quant ops.

Differential Revision: [D24479735](https://our.internmc.facebook.com/intern/diff/D24479735/)

[ghstack-poisoned]

Commit:
Update on "mem-efficient learnable fake quantization"


The updated quantization operator have a different gradient calculation for scale and zero_point when the output is at the endpoints of clamp operation. The updated quantization operator calculates the gradient according to the gradient of the `clamp` function. This behavior is consistent with the gradient calculation for non-learnable fake-quant ops.

Differential Revision: [D24479735](https://our.internmc.facebook.com/intern/diff/D24479735/)

[ghstack-poisoned]

Commit:
Update on "mem-efficient learnable fake quantization"


The updated quantization operator have a different gradient calculation for scale and zero_point when the output is at the endpoints of clamp operation. The updated quantization operator calculates the gradient according to the gradient of the `clamp` function. This behavior is consistent with the gradient calculation for non-learnable fake-quant ops.

Differential Revision: [D24479735](https://our.internmc.facebook.com/intern/diff/D24479735/)

[ghstack-poisoned]

Commit:
Update on "mem-efficient learnable fake quantization"


The updated quantization operator have a different gradient calculation for scale and zero_point when the output is at the endpoints of clamp operation. The updated quantization operator calculates the gradient according to the gradient of the `clamp` function. This behavior is consistent with the gradient calculation for non-learnable fake-quant ops.

Differential Revision: [D24479735](https://our.internmc.facebook.com/intern/diff/D24479735/)

[ghstack-poisoned]

Commit:
Update on "mem-efficient learnable fake quantization"


The updated quantization operator have a different gradient calculation for scale and zero_point when the output is at the endpoints of clamp operation. The updated quantization operator calculates the gradient according to the gradient of the `clamp` function. This behavior is consistent with the gradient calculation for non-learnable fake-quant ops.

Differential Revision: [D24479735](https://our.internmc.facebook.com/intern/diff/D24479735/)

[ghstack-poisoned]

Commit:
Update on "mem-efficient learnable fake quantization"


The updated quantization operator have a different gradient calculation for scale and zero_point when the output is at the endpoints of clamp operation. The updated quantization operator calculates the gradient according to the gradient of the `clamp` function. This behavior is consistent with the gradient calculation for non-learnable fake-quant ops.

Differential Revision: [D24479735](https://our.internmc.facebook.com/intern/diff/D24479735/)

[ghstack-poisoned]

Commit:
Update on "mem-efficient learnable fake quantization"


The updated quantization operator have a different gradient calculation for scale and zero_point when the output is at the endpoints of clamp operation. The updated quantization operator calculates the gradient according to the gradient of the `clamp` function. This behavior is consistent with the gradient calculation for non-learnable fake-quant ops.

Differential Revision: [D24479735](https://our.internmc.facebook.com/intern/diff/D24479735/)

[ghstack-poisoned]

Commit:
Update on "mem-efficient learnable fake quantization"


The updated quantization operator have a different gradient calculation for scale and zero_point when the output is at the endpoints of clamp operation. The updated quantization operator calculates the gradient according to the gradient of the `clamp` function. This behavior is consistent with the gradient calculation for non-learnable fake-quant ops.

Differential Revision: [D24479735](https://our.internmc.facebook.com/intern/diff/D24479735/)

[ghstack-poisoned]

Commit:
Update on "mem-efficient learnable fake quantization"


The updated quantization operator have a different gradient calculation for scale and zero_point when the output is at the endpoints of clamp operation. The updated quantization operator calculates the gradient according to the gradient of the `clamp` function. This behavior is consistent with the gradient calculation for non-learnable fake-quant ops.

Differential Revision: [D24479735](https://our.internmc.facebook.com/intern/diff/D24479735/)

[ghstack-poisoned]

Commit:
Update on "mem-efficient learnable fake quantization"


The updated quantization operator have a different gradient calculation for scale and zero_point when the output is at the endpoints of clamp operation. The updated quantization operator calculates the gradient according to the gradient of the `clamp` function. This behavior is consistent with the gradient calculation for non-learnable fake-quant ops.

Differential Revision: [D24479735](https://our.internmc.facebook.com/intern/diff/D24479735/)

[ghstack-poisoned]

Commit:
Update on "mem-efficient learnable fake quantization"


The updated quantization operator have a different gradient calculation for scale and zero_point when the output is at the endpoints of clamp operation. The updated quantization operator calculates the gradient according to the gradient of the `clamp` function. This behavior is consistent with the gradient calculation for non-learnable fake-quant ops.

Differential Revision: [D24479735](https://our.internmc.facebook.com/intern/diff/D24479735/)

[ghstack-poisoned]

Commit:
Update on "mem-efficient learnable fake quantization"


The updated quantization operator have a different gradient calculation for scale and zero_point when the output is at the endpoints of clamp operation. The updated quantization operator calculates the gradient according to the gradient of the `clamp` function. This behavior is consistent with the gradient calculation for non-learnable fake-quant ops.

Differential Revision: [D24479735](https://our.internmc.facebook.com/intern/diff/D24479735/)

[ghstack-poisoned]

Commit:
Update on "mem-efficient learnable fake quantization"


The updated quantization operator have a different gradient calculation for scale and zero_point when the output is at the endpoints of clamp operation. The updated quantization operator calculates the gradient according to the gradient of the `clamp` function. This behavior is consistent with the gradient calculation for non-learnable fake-quant ops.

Differential Revision: [D24479735](https://our.internmc.facebook.com/intern/diff/D24479735/)

[ghstack-poisoned]

Commit:
Update on "mem-efficient learnable fake quantization"


The updated quantization operator have a different gradient calculation for scale and zero_point when the output is at the endpoints of clamp operation. The updated quantization operator calculates the gradient according to the gradient of the `clamp` function. This behavior is consistent with the gradient calculation for non-learnable fake-quant ops.

Differential Revision: [D24479735](https://our.internmc.facebook.com/intern/diff/D24479735/)

[ghstack-poisoned]

Commit:
Update on "mem-efficient learnable fake quantization"


The updated quantization operator have a different gradient calculation for scale and zero_point when the output is at the endpoints of clamp operation. The updated quantization operator calculates the gradient according to the gradient of the `clamp` function. This behavior is consistent with the gradient calculation for non-learnable fake-quant ops.

Differential Revision: [D24479735](https://our.internmc.facebook.com/intern/diff/D24479735/)

[ghstack-poisoned]

Commit:
Update on "mem-efficient learnable fake quantization"


The updated quantization operator have a different gradient calculation for scale and zero_point when the output is at the endpoints of clamp operation. The updated quantization operator calculates the gradient according to the gradient of the `clamp` function. This behavior is consistent with the gradient calculation for non-learnable fake-quant ops.

Differential Revision: [D24479735](https://our.internmc.facebook.com/intern/diff/D24479735/)

[ghstack-poisoned]

walterddr commented:
Hi @haichuan-fb per our last discussion. I dont think mixing enabling this test and improving mem-efficiency should be in the same PR. it makes it hard to debug the CI issue. is it possible to first fix any issue that cause this test failure on `test_learnable_backward_per_channel_cuda` then add in the mem-efficient change?

haichuan-fb commented:
> Hi @haichuan-fb per our last discussion. I dont think mixing enabling this test and improving mem-efficiency should be in the same PR. it makes it hard to debug the CI issue. is it possible to first fix any issue that cause this test failure on `test_learnable_backward_per_channel_cuda` then add in the mem-efficient change?

Thanks. Yes, I agree.

Commit:
Update on "mem-efficient learnable fake quantization"


The updated quantization operator have a different gradient calculation for scale and zero_point when the output is at the endpoints of clamp operation. The updated quantization operator calculates the gradient according to the gradient of the `clamp` function. This behavior is consistent with the gradient calculation for non-learnable fake-quant ops.

Differential Revision: [D24479735](https://our.internmc.facebook.com/intern/diff/D24479735/)

[ghstack-poisoned]

raghuramank100 commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -2164,41 +2165,57 @@ void fake_quant_per_channel_cachemask_cpu(
 
 void fake_quantize_learnable_channel_grad_kernel_cpu(
-    TensorIterator& iter,
+    TensorIterator& iter_x,
```
Can you add a TODO stating that this needs to be optimized

raghuramank100 commented on test/quantization/test_workflow_module.py:
```diff
@@ -924,6 +925,7 @@ def test_backward_per_tensor_cachemask_cuda(self):
            X=hu.tensor(shapes=hu.array_shapes(1, 5,),
                        elements=hu.floats(-1e3, 1e3, allow_nan=False, allow_infinity=False),
                        qparams=hu.qparams(dtypes=torch.quint8)))
+    @unittest.skip("disable the test since py_module is replaced")
```
Why is the test disabled? With the call to the op, this test should still work.

raghuramank100 reviewed (approved):
Please enable disabled tests, otherwise this is good to land

Commit:
Update on "mem-efficient learnable fake quantization"


The updated quantization operator have a different gradient calculation for scale and zero_point when the output is at the endpoints of clamp operation. The updated quantization operator calculates the gradient according to the gradient of the `clamp` function. This behavior is consistent with the gradient calculation for non-learnable fake-quant ops.

Differential Revision: [D24479735](https://our.internmc.facebook.com/intern/diff/D24479735/)

[ghstack-poisoned]
